Assets/Scripts/MazeGenerator.cs

```diff
@@ -0,0 +1,164 @@
+
+    private readonly List<Vector2> offsets = new List<Vector2>
+    {
+        new Vector2(0, 1),
+        new Vector2(0, -1),
+        new Vector2(1, 0),
+        new Vector2(-1, 0)
+    };
+
+    private System.Random rnd = new System.Random();
+
+    private Vector2 currentTile;
+    public Vector2 CurrentTile
+    {
+        get { return currentTile; }
+
+        private set
+        {
+            if (value.x < 1 || value.x >= this.width - 1 || value.y < 1 || value.y >= this.height - 1)
+            {
+                throw new ArgumentException("CurrentTile must be within the one tile border all around the maze");
+            }
+            else if (value.x % 2 == 1 || value.y % 2 == 1)
+            {
+                currentTile = value;
+            }
+            else
+            {
+                throw new ArgumentException("The current square must not be both on an even X-axis and an even Y-axis, to ensure we can get walls around all tunnels");
+            }
+        }
+    }
+
+    private static MazeGenerator instance;
+    public static MazeGenerator Instance
+    {
+        get { return instance; }
+    }
+
+    private void Awake()
+    {
+        instance = this;
+        GenerateMaze();
+    }
+
+
+    private void GenerateMaze()
+    {
+        Maze = new int[width, height];
+        for (int x = 0; x < width; x++)
+        {
+            for (int y = 0; y < height; y++)
+            {
+                Maze[x, y] = (int)TileType.wall;
+            }
+        }
+
+        CurrentTile = Vector2.one;
+        tileStack.Push(CurrentTile);
+        PrimsAlg();
+
+        for (int i = 0; i <= Maze.GetUpperBound(0); i++)
+        {
+            for (int j = 0; j <= Maze.GetUpperBound(1); j++)
+            {
+                if (Maze[i, j] == (int)TileType.wall)
+                {
+                    Instantiate(brick, new Vector3(i * brick.localScale.x, 0, j * brick.localScale.z), Quaternion.identity);
+                }
+                else if (Maze[i, j] == (int)TileType.path)
+                {
+                    pathMazes.Add(new Vector3(i, 0, j));
+                }
+            }
+        }
+    }
+
+    private void PrimsAlg()
+    {
+        List<Vector2> neighbors;
+        while (tileStack.Count > 0)
+        {
+            Maze[(int)CurrentTile.x, (int)CurrentTile.y] = (int)TileType.path;
+
+            neighbors = GetValidNeighbors(CurrentTile);
+
+            if (neighbors.Count > 0)
+            {
+                tileStack.Push(CurrentTile);
+                CurrentTile = neighbors[rnd.Next(neighbors.Count)];
+            }
+            else
+            {
+                CurrentTile = tileStack.Pop();
+            }
+        }
+    }
+
+    /// <summary>
+    /// Get all the prospective neighboring tiles
+    /// </summary>
+    /// <param name="centerTile">The tile to test</param>
+    /// <returns>Any and all valid neighbors</returns>
+    private List<Vector2> GetValidNeighbors(Vector2 centerTile)
```

Comment:
This would be a lot better if you could accomplish it with out having offsets being a member of the class and instead something that is passed in